File: examples/dekita/.claude/hooks/existing-impl-check.py
#!/usr/bin/env python3
"""worktree作成時に既存実装の存在を警告し、検証を促す。

Why:
    「コードが存在する」≠「動作している」。Issueがオープンなのに
    関連コードが存在する場合、そのコードが正常に動作していない
    可能性が高い。検証せずに「実装済み」と判断すると時間を無駄にする。

What:
    - git worktree addコマンドからIssue番号を抽出
    - コメント・ファイル名からIssue番号やキーワードで関連コードを検索
    - 関連コードが見つかった場合は検証を促す警告を表示

Remarks:
    - ブロックせず警告のみ
    - 検索結果は最大5件まで表示

Changelog:
    - silenvx/dekita#????: フック追加
"""

import json
import re
import subprocess
import sys

from lib.constants import TIMEOUT_MEDIUM
from lib.execution import log_hook_execution
from lib.session import parse_hook_input


def extract_issue_number_from_command(command: str) -> int | None:
    """Extract issue number from git worktree add command.

    Checks both branch name (-b option) and path for issue numbers.
    """
    if "git worktree add" not in command:
        return None

    # Patterns to find issue numbers
    patterns = [
        r"issue[_-](\d+)",  # issue-123, issue_123
        r"#(\d+)",  # #123
        r"/(\d+)[-_]",  # /123-description
        r"[-_](\d+)[-_]",  # feature-123-name
        r"[-_](\d+)$",  # feature-123 (at end)
        r"[-_](\d+)\s",  # feature-123 (followed by space)
    ]

    for pattern in patterns:
        match = re.search(pattern, command, re.IGNORECASE)
        if match:
            return int(match.group(1))

    return None


def get_issue_title(issue_number: int) -> str | None:
    """Get issue title from GitHub."""
    try:
        result = subprocess.run(
            ["gh", "issue", "view", str(issue_number), "--json", "title"],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
        )
        if result.returncode == 0:
            data = json.loads(result.stdout)
            return data.get("title", "")
    except (subprocess.TimeoutExpired, json.JSONDecodeError, FileNotFoundError):
        pass  # Fail-open: エラー時はNoneを返す
    return None


def search_related_code(issue_number: int, issue_title: str | None) -> list[str]:
    """Search for code that might be related to this issue.

    Returns list of potentially related files.
    """
    related_files: list[str] = []

    # Search strategies:
    # 1. Search for issue number in code comments
    # 2. Search for keywords from issue title in filenames
    # 3. Search for common patterns based on issue type

    # Strategy 1: Issue number in comments (e.g., "# Issue #123", "// #123")
    try:
        result = subprocess.run(
            ["git", "grep", "-l", f"#{issue_number}"],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
        )
        if result.returncode == 0:
            related_files.extend(result.stdout.strip().split("\n"))
    except (subprocess.TimeoutExpired, FileNotFoundError):
        pass  # Fail-open: 検索失敗時は継続

    # Strategy 2: Extract keywords from issue title and search filenames
    if issue_title:
        # Extract potential function/file names from title
        # e.g., "feat(hooks): worktree作成時にIssue自動アサイン" → "auto-assign", "worktree"
        keywords = extract_keywords_from_title(issue_title)

        for keyword in keywords:
            if len(keyword) >= 4:  # Skip very short keywords
                try:
                    # Search for files containing the keyword in their name
                    result = subprocess.run(
                        ["git", "ls-files", f"*{keyword}*"],
                        capture_output=True,
                        text=True,
                        timeout=TIMEOUT_MEDIUM,
                    )
                    if result.returncode == 0 and result.stdout.strip():
                        for f in result.stdout.strip().split("\n"):
                            if f and f not in related_files:
                                related_files.append(f)
                except (subprocess.TimeoutExpired, FileNotFoundError):
                    pass  # Fail-open: 検索失敗時は継続

    # Deduplicate and filter
    seen = set()
    unique_files = []
    for f in related_files:
        # Filter out hidden files (like .gitignore) but keep .claude/ directory
        if f and f not in seen:
            if not f.startswith(".") or f.startswith(".claude/"):
                seen.add(f)
                unique_files.append(f)

    return unique_files[:5]  # Limit to 5 files


def extract_keywords_from_title(title: str) -> list[str]:
    """Extract potential code-related keywords from issue title."""
    keywords = []

    # Common patterns in issue titles
    # "feat(hooks): xxx" → extract "hooks"
    scope_match = re.search(r"\(([^)]+)\)", title)
    if scope_match:
        keywords.append(scope_match.group(1))

    # Extract hyphenated words that look like code names
    # e.g., "auto-assign", "issue-check"
    hyphenated = re.findall(r"[a-z]+-[a-z]+", title.lower())
    keywords.extend(hyphenated)

    # Extract camelCase or PascalCase words
    camel = re.findall(r"[A-Z][a-z]+[A-Z][a-z]+", title)
    keywords.extend([w.lower() for w in camel])

    # Japanese keywords that might indicate functionality
    jp_keywords = {
        "自動": "auto",
        "アサイン": "assign",
        "チェック": "check",
        "検証": "verif",
        "作成": "create",
        "削除": "delete",
        "更新": "update",
        "レビュー": "review",
        "マージ": "merge",
    }
    for jp, en in jp_keywords.items():
        if jp in title:
            keywords.append(en)

    return keywords


def main():
    """PreToolUse hook for Bash commands.

    Detects git worktree add and checks for existing implementations.
    """
    result = {"decision": "approve"}

    try:
        input_data = parse_hook_input()
        tool_input = input_data.get("tool_input", {})
        command = tool_input.get("command", "")

        # Only process git worktree add commands
        if "git worktree add" not in command:
            pass
        else:
            issue_number = extract_issue_number_from_command(command)
            if issue_number:
                issue_title = get_issue_title(issue_number)
                related_files = search_related_code(issue_number, issue_title)

                if related_files:
                    files_list = "\n".join(f"   - {f}" for f in related_files)
                    result["systemMessage"] = (
                        f"⚠️ **既存実装の検証が必要です** (Issue #{issue_number})\n\n"
                        f"関連する既存コードが見つかりました:\n{files_list}\n\n"
                        f"**重要**: 「コードが存在する」≠「動作している」\n"
                        f"Issueが存在する理由を確認し、既存実装が\n"
                        f"**実際に期待通り動作するか**を検証してください。\n\n"
                        f"検証せずに「実装済み」と判断すると、\n"
                        f"問題の見落としにつながります。"
                    )

    except Exception as e:
        # Don't block on errors
        print(f"[existing-impl-check] Error: {e}", file=sys.stderr)

    log_hook_execution(
        "existing-impl-check", result.get("decision", "approve"), result.get("reason")
    )
    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/existing_impl_check.ts
#!/usr/bin/env bun
/**
 * worktree作成時に既存実装の存在を警告し、検証を促す。
 *
 * Why:
 *   「コードが存在する」≠「動作している」。Issueがオープンなのに
 *   関連コードが存在する場合、そのコードが正常に動作していない
 *   可能性が高い。検証せずに「実装済み」と判断すると時間を無駄にする。
 *
 * What:
 *   - git worktree addコマンドからIssue番号を抽出
 *   - コメント・ファイル名からIssue番号やキーワードで関連コードを検索
 *   - 関連コードが見つかった場合は検証を促す警告を表示
 *
 * Remarks:
 *   - ブロックせず警告のみ
 *   - 検索結果は最大5件まで表示
 *
 * Changelog:
 *   - silenvx/dekita#????: フック追加
 *   - silenvx/dekita#2874: TypeScript移行
 */

import { spawnSync } from "node:child_process";
import { TIMEOUT_MEDIUM } from "../lib/constants";
import { formatError } from "../lib/format_error";
import { logHookExecution } from "../lib/logging";
import { parseHookInput } from "../lib/session";

const HOOK_NAME = "existing-impl-check";

/**
 * Extract issue number from git worktree add command.
 */
export function extractIssueNumberFromCommand(command: string): number | null {
  if (!command.includes("git worktree add")) {
    return null;
  }

  const patterns = [
    /issue[_-](\d+)/i, // issue-123, issue_123
    /#(\d+)/, // #123
    /\/(\d+)[-_]/, // /123-description
    /[-_](\d+)[-_]/, // feature-123-name
    /[-_](\d+)$/, // feature-123 (at end)
    /[-_](\d+)\s/, // feature-123 (followed by space)
  ];

  for (const pattern of patterns) {
    const match = pattern.exec(command);
    if (match) {
      return Number.parseInt(match[1], 10);
    }
  }

  return null;
}

/**
 * Get issue title from GitHub.
 */
function getIssueTitle(issueNumber: number): string | null {
  try {
    const result = spawnSync("gh", ["issue", "view", String(issueNumber), "--json", "title"], {
      encoding: "utf-8",
      timeout: TIMEOUT_MEDIUM * 1000,
    });

    if (result.status === 0) {
      const data = JSON.parse(result.stdout);
      return data.title || null;
    }
  } catch {
    // Fail-open
  }
  return null;
}

/**
 * Extract potential code-related keywords from issue title.
 */
export function extractKeywordsFromTitle(title: string): string[] {
  const keywords: string[] = [];

  // "feat(hooks): xxx" → extract "hooks"
  const scopeMatch = /\(([^)]+)\)/.exec(title);
  if (scopeMatch) {
    keywords.push(scopeMatch[1]);
  }

  // Extract hyphenated words that look like code names
  const hyphenated = title.toLowerCase().match(/[a-z]+-[a-z]+/g) || [];
  keywords.push(...hyphenated);

  // Extract camelCase or PascalCase words
  const camel = title.match(/[a-zA-Z][a-z]*[A-Z][a-z]*/g) || [];
  keywords.push(...camel.map((w) => w.toLowerCase()));

  // Japanese keywords that might indicate functionality
  const jpKeywords: Record<string, string> = {
    自動: "auto",
    アサイン: "assign",
    チェック: "check",
    検証: "verif",
    作成: "create",
    削除: "delete",
    更新: "update",
    レビュー: "review",
    マージ: "merge",
  };

  for (const [jp, en] of Object.entries(jpKeywords)) {
    if (title.includes(jp)) {
      keywords.push(en);
    }
  }

  return keywords;
}

/**
 * Search for code that might be related to this issue.
 */
function searchRelatedCode(issueNumber: number, issueTitle: string | null): string[] {
  const relatedFiles: string[] = [];

  // Strategy 1: Issue number in comments (e.g., "# Issue #123", "// #123")
  try {
    const result = spawnSync("git", ["grep", "-l", `#${issueNumber}`], {
      encoding: "utf-8",
      timeout: TIMEOUT_MEDIUM * 1000,
    });

    if (result.status === 0 && result.stdout.trim()) {
      relatedFiles.push(...result.stdout.trim().split("\n"));
    }
  } catch {
    // Fail-open
  }

  // Strategy 2: Extract keywords from issue title and search filenames
  if (issueTitle) {
    const keywords = extractKeywordsFromTitle(issueTitle);

    for (const keyword of keywords) {
      if (keyword.length >= 4) {
        try {
          const result = spawnSync("git", ["ls-files", `*${keyword}*`], {
            encoding: "utf-8",
            timeout: TIMEOUT_MEDIUM * 1000,
          });

          if (result.status === 0 && result.stdout.trim()) {
            for (const f of result.stdout.trim().split("\n")) {
              if (f && !relatedFiles.includes(f)) {
                relatedFiles.push(f);
              }
            }
          }
        } catch {
          // Fail-open
        }
      }
    }
  }

  // Deduplicate and filter
  const seen = new Set<string>();
  const uniqueFiles: string[] = [];

  for (const f of relatedFiles) {
    if (f && !seen.has(f)) {
      // Filter out hidden files but keep .claude/ directory
      if (!f.startsWith(".") || f.startsWith(".claude/")) {
        seen.add(f);
        uniqueFiles.push(f);
      }
    }
  }

  return uniqueFiles.slice(0, 5); // Limit to 5 files
}

interface HookResult {
  decision?: string;
  systemMessage?: string;
  reason?: string;
}

async function main(): Promise<void> {
  const result: HookResult = {};
  let sessionId: string | undefined;

  try {
    const inputData = await parseHookInput();
    sessionId = inputData.session_id;
    const toolInput = inputData.tool_input || {};
    const command = (toolInput as { command?: string }).command || "";

    // Only process git worktree add commands
    if (command.includes("git worktree add")) {
      const issueNumber = extractIssueNumberFromCommand(command);

      if (issueNumber) {
        const issueTitle = getIssueTitle(issueNumber);
        const relatedFiles = searchRelatedCode(issueNumber, issueTitle);

        if (relatedFiles.length > 0) {
          const filesList = relatedFiles.map((f) => `   - ${f}`).join("\n");
          result.systemMessage = `⚠️ **既存実装の検証が必要です** (Issue #${issueNumber})\n\n関連する既存コードが見つかりました:\n${filesList}\n\n**重要**: 「コードが存在する」≠「動作している」\nIssueが存在する理由を確認し、既存実装が\n**実際に期待通り動作するか**を検証してください。\n\n検証せずに「実装済み」と判断すると、\n問題の見落としにつながります。`;
        }
      }
    }
  } catch (e) {
    console.error(`[${HOOK_NAME}] Error: ${formatError(e)}`);
  }

  await logHookExecution(HOOK_NAME, "approve", result.reason, undefined, { sessionId });
  console.log(JSON.stringify(result));
}

if (import.meta.main) {
  main().catch((e) => {
    console.error(`[${HOOK_NAME}] Fatal error: ${formatError(e)}`);
    // Hooks should not block on internal errors - output approve response
    console.log(JSON.stringify({}));
    process.exit(0);
  });
}
